### ISSUE TITLE
Determine which serialisation / compression algorithm to use throughout the pipeline

### ISSUE THREAD

**Issue by ghost** [ooni/pipeline#40]
In order to reduce our data storage requirements, we should make use of compression as long as the benefits of doing so exceed any associated costs. To this end, we need to select a compression algorithm which weighs both the associated compression speed, and compression ratio.

To consider this issue complete, we need to identify:
- [ ] A compression algorithm to be used when submitting data to the collectors from a given ooni-probe;
- [ ] A compression algorithm to use throughout the pipeline when dealing with intermediary files; and
- [ ] A compression algorithm to use for long-term archiving

In certain cases, we may find the use of compression to be unnecessary (e.g. when dealing with intermediary I/O at each stage of the pipeline).

The top contenders right now are:
- pickle + lz4
- msgpack + zlib

I'll perform some A/B tests, and include my findings here.

@joelanders @hellais @darkk @willscott

**Comment by darkk** [comment]
> pickle

I'd rather avoid **storing** anything in a format tightly coupled with specific language unless it gives us major performance boost. It may be a good baseline, but I'm a bit sceptical about usage of pickle for data exchange in production.

**Comment by ghost** [comment]
@darkk I agree wholeheartedly. `pickle`/`dill` would only really be advantageous when working with metrics which haven't been normalised, and may not necessarily be JSON / BSON / msgpack serialisable.

**Comment by hellais** [comment]
I ran some benchmarks on the ooni-datacollector machine with the goal of identifying a good serialisation format for long term storage.

https://gist.github.com/hellais/c3914475331a313266c3f51e49bd287d

These results seem to indicate that lz4 could be a good option.

**Comment by darkk** [comment]
There are some more notes on compression and serialization I've gathered while passing some smoke through the pipeline.

tl;dr: `tar` small files — save IOPS, compress tarred — save bandwidth & space, read blocks not lines — `tarfile` line-by-line reading code is slow.

First, LZ4 [does not currently implement](https://github.com/lz4/lz4/blob/v1.7.4.2/lib/lz4frame.c#L873) checksums for **compressed** data (aka *Block checksum*), only content checksum is implemented, so we have to add some checksum that can be used for data scrubbing. I suggest to use crc32 due to following reasons:
- IMHO speedy checksum is better for scrubbing, I'm checksumming raw files with SHA1 and CRC32 to make scrubbing IO-bound instead of CPU-bound on every platform we can imagine
- the checksum should be easy to verify for others – xxhash, murmur cityhash don't have wide-spread CLI tools
- crc32 is not "strong", but lz4 also has xxhash content checksum, so I don't consider checksum "strength" as a requirement
- sha1 is a nice fallback for crc32, computation costs nothing as "canning" is bottlenecked by disk write

| box                             | disk | crc32 | sha512 | sha256 | sha1 | md5 |
| ------------------------------- | ----:| -----:| ------:| ------:| ----:| ---:|
| darkk-laptop i7-6600U, SSD      | 450  | 1120  | 576    | 317    | 846  | 568 |
| datacollector E5645, HDD? ceph? | 82   | 830   | 251    | 164    | 459  | 527 |
| ssdamsterdam  E5-2620, SSD      | 512  | 550   | 167    | 113    | 359  | 343 |

crc32 performance was estimated with `python2.7 -m timeit --number 100 --repeat 10 --setup  'import zlib; a = "A"*(1024*1024*16)' -- 'zlib.crc32(a)'`, others — with `openssl speed`.

Second, `python-lz4` does not implement streaming decompressor yet. That's not a big deal, we can use CLI `lz4` tool when we have to stream whole file. Sanitized files have small frames to be seekable, so streaming decompressor is required only for reading raw (*canned*) files.

Third, `lz4` is byte-oriented and not bit-oriented, so it does not compress random base64-encoded strings well, it means, that we MAY want to have different compression codes for different reports if/when amount of binary base64-encoded data rises (pcaps, TLS certificates, non-unicode text of the webpages, and so on).

```
$ cat urandom.1Mb | base64 | wc --bytes
1416501
$ cat urandom.1Mb | base64 | gzip -6 | wc --bytes
1076437
$ cat urandom.1Mb | base64 | lz4 -6 | wc --bytes
1416520
```

Fourth, legendary slowness of `yaml` processing has two causes, `yaml` library is not the only one. `yaml.loader.CLoader` processes ~40 Mbytes of raw yamly reports per second, BUT reading tons of tiny yaml files with simple `cat *.yaml` is way slower than 'usual' disk speed, it was less than 10 Mbyte/s on CEPH-based disk. Compressing several yaml files into tar archive and streaming it to the CLoader works significantly better being bottlenecked by CPU. Mobile app will produce alike pattern as it was going to have quite small reports.
`tarfile` itself does not have major overhead, stream decoding speed is ~450 Mbyte/s at datacollector and ~530 at darkk-laptop (given cached tarfile as an input). IO-related measurements were done with `vm.drop_caches=3` to ensure that the data is not read from local pagecache.

**Comment by hellais** [comment]
I believe we have now settled on using a tar container compressed using lz4, so I think this can now be closed.

@darkk am I correct?

**Comment by darkk** [comment]
Yep.
And `.tar.{gz|bz2|xz}` for cold storage in glacier as `lz4` is not settled yet -- TheTorProject/ooni-pipeline#57
And adding `gzip` to collector, ooni-probe and MK (zlib is part of iOS SDK and Android NDK) as every report is POSTed line-by-line, uncompressed (probably, that deserves some batching and/or keepalive) -- TheTorProject/ooni-backend#106.
